presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrderCurveSearcher.java

```diff
@@ -114,21 +129,79 @@ public static ZOrderCurveSearcher createZOrderCurveSearcher(ZOrder zOrder, List<
         for (int index = 0; index < valueRanges.size(); index++) {
             ZValueRange range = valueRanges.get(index);
 
-            Integer minimumValue = range.getMinimumValue();
-            Integer maximumValue = range.getMaximumValue();
-
-            if (minimumValue == null) {
-                minimumValue = 0;
+            List<Integer> minimumValues = range.getMinimumValues();
+            List<Optional<Integer>> optionalMinimumValues = new ArrayList<>();
+            for (Integer minimumValue : minimumValues) {
+                if (minimumValue == null) {
+                    minimumValue = 0;
+                }
+                optionalMinimumValues.add(Optional.of(minimumValue));
             }
-            if (maximumValue == null) {
-                maximumValue = (1 << encodingBits.get(index)) - 1;
+
+            List<Integer> maximumValues = range.getMaximumValues();
+            List<Optional<Integer>> optionalMaximumValues = new ArrayList<>();
+            for (Integer maximumValue : maximumValues) {
+                if (maximumValue == null) {
+                    maximumValue = (1 << encodingBits.get(index)) - 1;
+                }
+                optionalMaximumValues.add(Optional.of(maximumValue));
             }
 
-            nonNullValues.add(new ZValueRange(Optional.of(minimumValue), Optional.of(maximumValue)));
+            nonNullValues.add(new ZValueRange(optionalMinimumValues, optionalMaximumValues));
```

Comment:
This is a thought, but I'm wondering if it makes sense to have a `BoundedZValueRange` which uses unboxed type so it rules out the possibility of null. Currently it is possible to have null/empty value for bounds in `ZValueRange`, but after null-filling, they will become bounded, and I guess we only operate on the bounded range?

Reply:
Good idea! I was considering doing something similar, but didn't think it all the way through. Thanks for this suggestion!